src/components/interface/cbuf_mgr/cos_alloc.c

```diff
@@ -0,0 +1,536 @@
+#define MAP_FAILED ((void*)0)  //((void*)-1)
+#endif
+
+#ifndef NULL
+#define NULL ((void*)0)
+#endif
+
+typedef struct {
+  cbuf_t cbuf_id;
+  void*  next;
+  size_t size;
+} __alloc_t;
+
+#define BLOCK_START(b)	(((void*)(b))-sizeof(__alloc_t))
+#define BLOCK_RET(b)	(((void*)(b))+sizeof(__alloc_t))
+
+#define MEM_BLOCK_SIZE	PAGE_SIZE
+#define PAGE_ALIGN(s)	round_up_to_page(s)
+
+#ifdef NIL
+#define REGPARM(x) __attribute__((regparm(x)))
+#else
+#define REGPARM(x)
+#endif
+
+#ifdef USE_VALLOC
+void *cos_get_vas_page(void)
+{
+	return valloc_alloc(cos_spd_id(), cos_spd_id(), 1);
+}
+
+void cos_release_vas_page(void *p)
+{
+	valloc_free(cos_spd_id(), cos_spd_id(), p, 1);
+}
+#endif
+
+/* -- SMALL MEM ----------------------------------------------------------- */
+
+static __alloc_t* __small_mem[8];
+
+#define __SMALL_NR(i)		(MEM_BLOCK_SIZE/(i))
+
+#define __MIN_SMALL_SIZE	__SMALL_NR(256)		/*   16 /   32 */
+#define __MAX_SMALL_SIZE	__SMALL_NR(2)		/* 2048 / 4096 */
+
+#define GET_SIZE(s)		(__MIN_SMALL_SIZE<<get_index((s)))
+
+#define FIRST_SMALL(p)		(((unsigned long)(p))&(~(MEM_BLOCK_SIZE-1)))
+
+static inline int __ind_shift() { return (MEM_BLOCK_SIZE==4096)?4:5; }
+
+static size_t REGPARM(1) get_index(size_t _size) {
+	size_t idx, size;
+	for (idx = 0, size = ((_size-1)&(MEM_BLOCK_SIZE-1))>>__ind_shift() ;  
+	     size ; 
+	     size>>=1, ++idx );
+	return idx;
+}
+
+#if ALLOC_DEBUG >= ALLOC_DEBUG_STATS
+struct mem_stats {
+	unsigned long long alloc, free;
+};
+typedef enum { DBG_ALLOC, DBG_FREE } alloc_dbg_type_t;
+static struct mem_stats __mem_stats[8];
+
+void alloc_stats_print(void)
+{
+	int i;
+
+	for (i = 0 ; i < 8 ; i++) {
+		unsigned long long min;
+
+		printc("cos_alloc: bin %d, alloc %lld, free %lld", 
+		       i, __mem_stats[i].alloc, __mem_stats[i].free);
+		min = (__mem_stats[i].alloc < __mem_stats[i].free) ? 
+			__mem_stats[i].alloc : 
+			__mem_stats[i].free;
+		__mem_stats[i].free  -= min;
+		__mem_stats[i].alloc -= min;
+	}
+}
+
+static void alloc_stats_report(alloc_dbg_type_t type, int bin)
+{
+	switch (type) {
+	case DBG_ALLOC:
+		__mem_stats[bin].alloc++;
+		break;
+	case DBG_FREE:
+		__mem_stats[bin].free++;
+		break;
+	}
+}
+#else
+#define alloc_stats_report(t, b)
+#endif
+
+/* small mem */
+static void __small_free(void*_ptr,size_t _size) REGPARM(2);
+
+static inline void REGPARM(2) __small_free(void*_ptr,size_t _size) {
+	__alloc_t* ptr=BLOCK_START(_ptr), *prev;
+	size_t size=_size;
+	size_t idx=get_index(size);
+	
+//	memset(ptr,0,size);	/* allways zero out small mem */
+#if ALLOC_DEBUG >= ALLOC_DEBUG_ALL
+	if (alloc_debug) printc("free (in %d): freeing %p of size %d and index %d.", 
+				cos_spd_id(), ptr, size, idx);
+#endif
+	do {
+		prev = __small_mem[idx];
+		ptr->next=prev;
+	} while (unlikely(cos_cmpxchg(&__small_mem[idx], (int)prev, (int)ptr) != (int)ptr));
+	alloc_stats_report(DBG_FREE, idx);
+}
+
+static void* do_cbuf_alloc(size_t size, cbuf_t *cbid)
+{
+    if (!size)
+    {
+        return NULL;
```

Comment:
I would just put these all on the same line `if (!size) return NULL` as it is an error case, and you want to de-emphasize it in the code.